dcpandas/df_builder.py

```diff
@@ -0,0 +1,292 @@
+# Copyright 2020 Google Inc.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#   http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""Data Commons Pandas API DataFrame Builder Module.
+
+Provides functions for building pandas DataFrames using the Data Commons Graph.
+"""
+
+from __future__ import absolute_import
+from __future__ import division
+from __future__ import print_function
+
+import collections
+import pandas as pd
+import six
+
+import datacommons.stat_vars as dc
+
+
+def build_time_series(place, stat_var):
+    """Constructs a pandas Series with `dates` as the index and corresponding `stat_var` statistics as values.
+    
+    Args:
+      place (`str`): The dcid of Place to query for.
+      stat_var (`str`): The dcid of the StatisticalVariable.
+    Returns:
+      A pandas Series with Place IDs as the index, and Observed statistics as values.
+    """
+    return pd.Series(dc.get_stat_series(place, stat_var))
+
+
+def _group_stat_all_by_obs_options(places, stat_vars, mode):
+    """Groups the result of `get_stat_all` by Observation options for time series.
+    
+    Args:
+      places (`str` or `iterable` of `str`): The dcids of Places to query for.
+      stat_vars (`Iterable` of `str`): The dcids of the StatisticalVariables.
+      mode (`str`): "series" to output time series grouped by Observation options, or
+        "covariates" to output latest Observations.
+    Returns:
+      A pandas Series with Place IDs as the index, and Observed statistics as values.
+
+    Raises:
+      ValueError: If the payload returned by the Data Commons REST API is
+        malformed.
+    """
+    kseries = "series"
+    kcov = "covariates"
+
+    if mode == kseries:
+        if len(stat_vars) != 1:
+            raise ValueError(
+                'When `mode=series`, only one StatisticalVariable for `stat_vars` is allowed.'
+            )
+        res = collections.defaultdict(list)
+    elif mode == kcov:
+        res = collections.defaultdict(lambda: collections.defaultdict(list))
+    else:
+        raise ValueError(
+            'Value of `mode` must be one of ("series", "covariates")')
+
+    stat_all = dc.get_stat_all(places, stat_vars)
+    for place, place_data in stat_all.items():
+        if not place_data:
+            continue
+        for stat_var, stat_var_data in place_data.items():
+            if not stat_var_data:
+                continue
```

Comment:
I added this to the docstring:

```
    Note that this function does not preserve `(place, stat_var)` pairs that
    yield no data `from get_stat_all`. In the extreme case that there is no
    data for any pairs, raise a ValueError instead of returning an empty dict.
```
 
and  updated the Raises section of the docstring

```
    Raises:
      ValueError: If the payload returned by the Data Commons REST API is
        malformed, or if there is no data for any (Place, StatisticalVariables)
        pair.
```


I decided to do the raise here (closest to source of issue) instead of in the dataframe public functions. Also fixed our tests to be able to test this in df_builder_test. We should now be able to add more assertRaises in our testing. Sorry for not figuring out the issue earlier.